aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt

```diff
@@ -256,4 +262,83 @@ class RealAWSCognitoAuthPluginTest {
         // THEN
         coVerify { anyConstructed<ResetPasswordUseCase>().execute(username, options, onSuccess, onError) }
     }
+
+    @Test
+    fun `fetch user attributes with success`() {
+        // GIVEN
+        val onSuccess = mockk<Consumer<MutableList<AuthUserAttribute>>>(relaxed = true)
+        val onError = mockk<Consumer<AuthException>>(relaxed = true)
+
+        val currentAuthState = mockk<AuthState> {
+            every { authNState } returns AuthenticationState.SignedIn(mockk())
+        }
+        every { authStateMachine.getCurrentState(captureLambda()) } answers {
+            lambda<(AuthState) -> Unit>().invoke(currentAuthState)
+        }
+
+        val credential = AmplifyCredential(
+            CognitoUserPoolTokens("idToken", "accessToken", "refreshToken", 120L),
+            null,
+            null
+        )
+
+        val eventSlot = CapturingSlot<(CredentialStoreState) -> Unit>()
+        every { credentialStoreStateMachine.listen(capture(eventSlot), any()) } answers {
+            eventSlot.captured.invoke(CredentialStoreState.Success(credential))
+            UUID.randomUUID()
+        }
+
+        val userAttributes = listOf(
+            AttributeType.invoke {
+                name = "email"
+                value = "email"
+            },
+            AttributeType.invoke {
+                name = "nickname"
+                value = "nickname"
+            }
+        )
+
+        coEvery {
+            authService.cognitoIdentityProviderClient?.getUser(any<GetUserRequest>())
+        } returns GetUserResponse.invoke {
+            this.userAttributes = userAttributes
+        }
+
+        // WHEN
+        plugin.fetchUserAttributes(onSuccess, onError)
+        Thread.sleep(1_000)
+        assertTrue { eventSlot.isCaptured }
+
+        val expectedResult = buildList {
+            userAttributes.mapTo(this) {
+                AuthUserAttribute(
+                    AuthUserAttributeKey.custom(it.name),
+                    it.value
+                )
+            }
+        }
+        verify { onSuccess.accept(expectedResult.toMutableList()) }
+        coVerify(exactly = 0) { onError.accept(any()) }
+    }
+
+    @Test
+    fun `fetch user attributes fails when not in SignedIn state`() {
+        // GIVEN
+        val onSuccess = mockk<Consumer<MutableList<AuthUserAttribute>>>(relaxed = true)
+        val onError = mockk<Consumer<AuthException>>(relaxed = true)
+
+        val currentAuthState = mockk<AuthState> {
+            every { authNState } returns AuthenticationState.NotConfigured()
+        }
+        every { authStateMachine.getCurrentState(captureLambda()) } answers {
+            lambda<(AuthState) -> Unit>().invoke(currentAuthState)
+        }
+        // WHEN
+        plugin.fetchUserAttributes(onSuccess, onError)
+        Thread.sleep(1_000)
```

Comment:
Do we need to sleep for so long? This can compound and cause issues in our overall testing time?

Reply:
agreed, we should not call sleep. Instead, I suggest to use something like countdown latch with timeout